plugin/keyring_vault/vault_keyring.cc

```diff
@@ -0,0 +1,219 @@
+    IKeyring_io *keyring_io(new Vault_io(logger.get(), vault_curl.release(), vault_parser.release()));
+    if (new_keys->init(keyring_io, keyring_filename))
+    {
+      mysql_rwlock_unlock(&LOCK_keyring);
+      return 1;
+    }
+    *reinterpret_cast<IKeys_container **>(save)= new_keys.release();
+    mysql_rwlock_unlock(&LOCK_keyring);
+  }
+  catch (const std::bad_alloc &e)
+  {
+    DBUG_ASSERT(0);
+    if (logger != NULL)
+      logger->log(MY_ERROR_LEVEL, "Cannot set keyring_vault_config_file "
+                                  "due to memory allocation failure");
+    mysql_rwlock_unlock(&LOCK_keyring);
+    return TRUE;
+  }
+  catch (...)
+  {
+    DBUG_ASSERT(0);
+    if (logger != NULL)
+      logger->log(MY_ERROR_LEVEL, "Cannot set keyring_vault_config_file due to internal"
+                                  "exception inside the keyring_vault plugin");
+    mysql_rwlock_unlock(&LOCK_keyring);
+    return 1;
+  }
+  return(0);
+}
+
+static char *keyring_vault_config_file= NULL;
+static MYSQL_SYSVAR_STR(
+  config,                                                      /* name       */
+  keyring_vault_config_file,                                   /* value      */
+  PLUGIN_VAR_RQCMDARG,                                         /* flags      */
+  "The path to the keyring_vault configuration file",          /* comment    */
+  check_keyring_file_data,                                     /* check()    */
+  update_keyring_file_data,                                    /* update()   */
+  ""                                                           /* default    */
+);
+
+static struct st_mysql_sys_var *keyring_vault_system_variables[]= {
+  MYSQL_SYSVAR(config),
+  NULL
+};
+
+static int keyring_vault_init(MYSQL_PLUGIN plugin_info)
+{
+  try
+  {
+#ifdef HAVE_PSI_INTERFACE
+    keyring_init_psi_keys();
+#endif
+    if (init_keyring_locks())
+      return TRUE;
+
+    logger.reset(new Logger(plugin_info));
+    keys.reset(new Vault_keys_container(logger.get()));
+    boost::movelib::unique_ptr<IVault_curl> vault_curl(new Vault_curl(logger.get()));
+    boost::movelib::unique_ptr<IVault_parser> vault_parser(new Vault_parser(logger.get()));
+    IKeyring_io *keyring_io= new Vault_io(logger.get(), vault_curl.release(),
+                                          vault_parser.release());
+    if (keys->init(keyring_io, keyring_vault_config_file))
+    {
+      is_keys_container_initialized = FALSE;
+      logger->log(MY_ERROR_LEVEL, "keyring_vault initialization failure. Please check that"
+        " the keyring_vault_config_file points to readable keyring_vault configuration"
+        " file. Please also make sure Vault is running and accessible."
+        " The keyring_vault will stay unusable until correct configuration file gets"
+        " provided.");
+      return FALSE;
+    }
+    is_keys_container_initialized = TRUE;
+    return FALSE;
+  }
+  catch (const std::bad_alloc &e)
+  {
+    DBUG_ASSERT(0);
```

Comment:
I did add a helper function, but I am not sure e.what() is a good idea. We have no control over what information will be returned in what - thus we cannot guarantee that sensitive information will not be included in the log message. Also I think DBUG_ASSERT(0) should stay - since we "eat" an exception so it would no go outside the plugin in to the server - this is the only way the debug server would crash with a stack trace that we could examine.